### PR TITLE
Explorer: Support class subtyping

### PR DESCRIPTION
**Pull request by Pixep**
Features:
* Adds support for [subtyping](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/classes.md#subtyping) for local variables, and function parameters

Changes:
* Update function parameter `Deduce` to handle subtyping
* Update `InstantiateType` to support `PointerType`
* Update `Convert` to support convertion from child class to a base class

Relates to #1881

### CODE REVIEW

File: explorer/testdata/class/class_subtyping_fn_param.carbon
Comment (jonmeow) [inline explorer/testdata/class/class_subtyping_fn_param.carbon]: With this change, is there a test for a pointer conversion that fails? e.g., dropping `extends` from this test. I think it may be good to keep one, just to be safe.

File: explorer/interpreter/value.cpp
Comment (jonmeow) [inline explorer/interpreter/value.cpp]:
```suggestion
  std::optional<Nonnull<const NominalClassType*>> ancestor_class = this;
  while (ancestor_class) {
    if (TypeEqual(ancestor_class.value(), other_class, std::nullopt)) {
      return true;
    }
    ancestor_class = ancestor_class.value()->base();
```

I think `ancestor` better communicates the use of the cursor. Suggesting `ancestor_class` just to mirror `other_class`.

File: explorer/interpreter/interpreter.cpp
Comment (jonmeow) [inline explorer/interpreter/interpreter.cpp]:
```suggestion
      // parameterized types for pointers in function call parameters.
```

File: explorer/interpreter/interpreter.cpp
Comment (jonmeow) [inline explorer/interpreter/interpreter.cpp]:
```suggestion
        // No conversion needed.
```

File: explorer/interpreter/interpreter.cpp
Comment (jonmeow) [inline explorer/interpreter/interpreter.cpp]:
```suggestion
      // Get pointed value.
```

File: explorer/interpreter/interpreter.cpp
Comment (jonmeow) [inline explorer/interpreter/interpreter.cpp]: Is this a question because you're not sure and think it won't occur? If so, you might instead CARBON_CHECK -- i.e., if this code shouldn't run, add a CHECK and if someone figures out code that causes it, then it's done.

File: explorer/interpreter/interpreter.cpp
Comment (jonmeow) [inline explorer/interpreter/interpreter.cpp:873]: I'm trying to understand this. The value here is some memory address. The destination type is, say, i32. The conversion result is the memory address, i.e. success?

I think this needs to be tested a bit more; do we have things like:

```
var a: A = {};
var b: A* = &a;
var c: i32 = b;
```

```
var a: i32 = 0;
var b: i32* = &a;
var c: A* = b;
```

```
var a: A = {};
var b: A* = &a;
var c: A** = b;
```

```
var a: A = {};
var b: A* = &a;
var c: A* = &b;
```

(all these tests are written as fail-tests)

Comment (Pixep) [reply explorer/interpreter/interpreter.cpp:873]: I should be a bit more explicit: Whether or not we execute `Convert` today is mainly driven by `IsImplicitlyConvertible`, and the previous code was returning the value as-is in _all_ situations. Trying to constrain that more led to some of our test cases failing, mainly around generics, because `Convert` is called with pointer types that are not always fully substituted, and I couldn't find a simple way around that, that would fit in the PR.

So yes, we are allowing any and all conversion from the `Convert` function point of view, but in actuality is limited by `IsImplicitlyConvertible`. I can add a TODO to track this gap.

File: explorer/interpreter/interpreter.cpp
Comment (Pixep) [reply explorer/interpreter/interpreter.cpp]: The question is a real question actually :D, I was hoping you or @zygoloid might know. I saw it happen, and don't understand what this represents (maybe during type-checking only?). Killing this branch leads to errors. Either way, something to figure out before merging.

File: explorer/testdata/class/class_subtyping_fn_param.carbon
Comment (Pixep) [reply explorer/testdata/class/class_subtyping_fn_param.carbon]: Good call
Done.

File: explorer/interpreter/interpreter.cpp
Comment (jonmeow) [reply explorer/interpreter/interpreter.cpp:873]: So, I think there are two possibilities here:
1) The code path is unused, CARBON_CHECK is okay
2) The code started running

I think the suggested tests currently fail, e.g.:
https://carbon.compiler-explorer.com/z/s9ecvezfe
https://carbon.compiler-explorer.com/z/qe64jEbYr

So let's add the tests explicitly, and make sure there's no regression.

File: explorer/interpreter/interpreter.cpp
Comment (jonmeow) [reply explorer/interpreter/interpreter.cpp]: Without seeing the errors, I would expect this branch needs to be adjusted. What's the destination_type?

This might be a type checking thing, i.e. `A*` being PointerType -> NominalClassType. But It should probably still depend on what the destination_type is: destination_type there is probably either "Type" or "PointerType -> Type".

File: explorer/interpreter/interpreter.cpp
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]:
```suggestion
      // Get pointee value.
```

File: explorer/interpreter/interpreter.cpp
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]: I think we should prefer to use the pointer-like syntax for optionals where practical, because it's more concise, and makes it easier to distinguish the application logic from the mechanics of wrapping and unwrapping. It's also probably more consistent with how the corresponding Carbon code will look, and experimenting with writing Carbon-like code is the whole point of using `optional<Nonnull<T*>>` instead of `T*`.

```suggestion
        if (TypeEqual(&(*class_subobj)->type(), &dest_ptr->type(),
```

File: explorer/interpreter/interpreter.cpp
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]:
```suggestion
        class_subobj = (*class_subobj)->base();
```

File: explorer/interpreter/interpreter.cpp
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]: Would it work to use `destination_type` instead of `dest_ptr->type()` here? I think we should be cautious about accessing the type of a run-time `Value` in the interpreter, because the compiler won't be able to do that.

File: explorer/interpreter/type_checker.cpp
Comment (geoffromer) [inline explorer/interpreter/type_checker.cpp]: We should probably be consistent about whether we say "pointed" or "pointee". I prefer "pointee" because it's clearly a noun, whereas "pointed" could be read as a verb (e.g. maybe `param_pointed` is a `bool` that indicates whether a parameter pointed to something).

File: explorer/interpreter/value.cpp
Comment (geoffromer) [inline explorer/interpreter/value.cpp]:
```suggestion
    if (TypeEqual(*ancestor_class, other_class, std::nullopt)) {
```

File: explorer/interpreter/value.cpp
Comment (geoffromer) [inline explorer/interpreter/value.cpp]:
```suggestion
    ancestor_class = (*ancestor_class)->base();
```

File: explorer/testdata/class/class_subtyping.carbon
Comment (geoffromer) [inline explorer/testdata/class/class_subtyping.carbon:1]: It seems like the main thing this is testing is non-virtual dispatch, rather than class subtyping more broadly, so maybe it should be named something like `non_virtual_dispatch.carbon`?

**Review by geoffromer — COMMENTED** [review]
(Migrating comments from #2462)

File: explorer/interpreter/interpreter.cpp
Comment (Pixep) [reply explorer/interpreter/interpreter.cpp:873]: Tests added and passing ✔️

File: explorer/interpreter/interpreter.cpp
Comment (zygoloid) [inline explorer/interpreter/interpreter.cpp]: This seems like too much "working out the semantics" to be doing in the interpreter -- I think the type-checker should instead work out how many levels of `.base` are implied by the implicit conversion, and add that information to the AST. Perhaps we should rewrite an implicit derived-to-base conversion into a series of `.base` member accesses?

File: explorer/interpreter/type_checker.cpp
Comment (zygoloid) [inline explorer/interpreter/type_checker.cpp]: This should only be permitted when `allow_implicit_conversion` is `true`.

File: explorer/interpreter/type_checker.cpp
Comment (zygoloid) [inline explorer/interpreter/type_checker.cpp]: We'll also need to decide if we want a C++-like deduce-from-base-class rule:
```
base class B(T:! Type) {}
class D extends B(i32) {}
fn F[T:! Type](p: B(T)*) {}
fn Main() {
  var d: D = {};
  // Is this OK, deducing T = i32, or an error?
  F(&d);
}
```
I've filed this as #2464. I'm happy for us to disallow this for now (as this change does), but I think it'd be useful to add a TODO mentioning that issue and the above testcase or something like it.

File: explorer/interpreter/type_checker.cpp
Comment (Pixep) [reply explorer/interpreter/type_checker.cpp]: Makes sense 👍 , the rest of the code already uses pointee

File: explorer/interpreter/type_checker.cpp
Comment (Pixep) [reply explorer/interpreter/type_checker.cpp]: Good point, done.

File: explorer/interpreter/type_checker.cpp
Comment (Pixep) [reply explorer/interpreter/type_checker.cpp]: Added

File: explorer/testdata/class/class_subtyping.carbon
Comment (Pixep) [reply explorer/testdata/class/class_subtyping.carbon:1]: Consolidated the subtyping part with the other, and renamed

File: explorer/interpreter/interpreter.cpp
Comment (Pixep) [reply explorer/interpreter/interpreter.cpp]: Tested the branch again, and it isn't used anymore, I think this was caused by `BaseElement` previously returning a different type. Fixed, thanks for suggestions.

File: explorer/interpreter/interpreter.cpp
Comment (Pixep) [reply explorer/interpreter/interpreter.cpp]: Not really unfortunately, `destination_type` would just tell us it is a pointer in general, but we need to match the pointee type with the destination pointer type.

Comment (jonmeow) [reply explorer/interpreter/interpreter.cpp]: To try to figure out what this might look like -- are you suggesting something like adding rewrite support for all expressions, and inserting "OperatorExpression(As)" for each incremental step?

e.g., var `x: D* = A.MakePointer()` becomes `x: D* = A.MakePointer() as B* as C* as D*`, adding recursive rewrites onto the expression node that represents the initializer?

Note, I don't think I have a strong opinion either way -- we could state that implicit casts only work for one step, or that implicit casts work for the full reach. I'm not sure how much moving the logic to the type checker would be cleaner versus just hiding that this is still doing something similar (since in either case, this code assumes the type checker has already figured out the indicated cast is feasible).

Comment (geoffromer) [reply explorer/interpreter/interpreter.cpp]: Ah, sorry, I misunderstood what this code was doing. In my defense, `PointerType::type` is very confusingly-named; I've sent #2468 to fix that.

Comment (zygoloid) [reply explorer/interpreter/interpreter.cpp]: As a concrete example of how I'm imagining we might model this:

- Change `BuiltinConvertExpr` to be rewritable, following the same pattern as the other expression forms that derive from `RewritableMixin<T>`.
- When the type-checker performs an implicit conversion from derived to base, traversing N levels of class hierarchy, it rewrites that implicit conversion into `operand.base.base.base` where there are N levels of `.base`. Either:
    - Each level is represented as a `SimpleMemberAccessExpression` whose `member` is a `BaseElement`. This would mean that the `member` of a `SimpleMemberAccessExpression` changes from being a `NamedElement` to being any `Element`. Or,
    - Each level is represented by some new form of expression that performs a member access of a `BaseElement.

Note that we should already be building a `BuiltinConvertExpr` for every implicit conversion that uses one of the built-in rules, and we already build a call to `ImplicitAs.Convert` for every implicit conversion that uses a user-defined `impl`, so all the contexts in which an implicit conversion can happen already support having their operands rewritten; that wouldn't be new here.

Comment (Pixep) [reply explorer/interpreter/interpreter.cpp]: Thank you both.

I agree with Jon that I don't see it making a lot of difference today, because the fact that we try to convert here, means that there _is_ a valid subtyping conversion. I can rewrite the surrounding code to highlight that with `CARBON_FATAL` or `CHECK`.

But it would make more sense from a design standpoint for the interpreter to already have that information available yes. So same as Jon, no strong opinion.

I started looking at refactoring based on the example provided, thanks a lot!

Comment (zygoloid) [reply explorer/interpreter/interpreter.cpp]: I'm happy for this to be changed to do more of the work in the type-checker and less in the interpreter in a follow-up pull request. And I'm going to be on vacation for a couple of weeks; so please don't block the progress of this patch on me or this comment thread!

Comment (Pixep) [reply explorer/interpreter/interpreter.cpp]: I think it'd make sense to merge this one first then. The suggested change is in progress in parallel, and that'd unblock the #2442 as well. Would that work for you @jonmeow ?

**Review by jonmeow — APPROVED** [review]

File: explorer/interpreter/interpreter.cpp
Comment (jonmeow) [reply explorer/interpreter/interpreter.cpp]: I'm good with that plan; reviewed, approved and merging.

**Pull request merged**
